### PR TITLE
make typeFilter accept arbitrary sets of OSMType

### PR DESCRIPTION
**Pull request by tyrasd**
This changes osm-type filters to

* make them work the same as the other filters (i.e. when called multiple times, the filters are intersected together). technically this is a breaking change compared to how this worked
* allow arbitrary set types to be supplied (not only `EnumSet`s)
* made jdbc backends work with "no osm type" cases. (e.g. `.osmType(EnumSet.noneOf(OSMType.class))`).
* add unit tests

# Checklist
- [x] My code follows the [code-style](https://github.com/GIScience/oshdb/blob/master/CONTRIBUTING.md) rules and I have checked on the [static analyses](https://jenkins.ohsome.org/job/oshdb/view/change-requests/) and [benchmark](https://reports.ohsome.org/oshdb-benchmarks/) (if applicable) results
- [x] I have commented my code
- [x] I have written javadoc (required for public methods)
- [x] I have added sufficient unit tests
- [x] I have made corresponding changes to the [documentation](https://github.com/GIScience/oshdb/tree/master/documentation)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/oshdb/blob/master/CHANGELOG.md)
- [x] I have adjusted the [examples](https://gitlab.gistools.geog.uni-heidelberg.de/giscience/big-data/ohsome/oshdb-examples) and [benchmark](https://reports.ohsome.org/oshdb-benchmarks/) if necessary

### CODE REVIEW

File: oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java [hunk 1]
```diff
@@ -80,6 +82,52 @@ public void multiPolygon() throws Exception {
...
+    assertTrue(result.equals(EnumSet.noneOf(OSMType.class)));
+    // called multiple times
+    result = createMapReducerOSMEntitySnapshot()
+        .osmType(OSMType.NODE)
+        .osmType(EnumSet.allOf(OSMType.class))
+        .areaOfInterest(bbox)
+        .timestamps(timestamps1)
+        .map(snapshot -> snapshot.getEntity().getType())
+        .stream().collect(Collectors.toSet());
+    assertTrue(result.equals(EnumSet.of(OSMType.NODE)));
```
[the remark this answers is not in the document]
Comment (FabiKo117) [reply oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java]: Ok I read through it again and this implementation seems to be fine.
The description at the beginning with the 'and combination' is a bit misleading imo as here the 2nd call is basically just ignored. I guess this Sets.intersetion() means basically that e.g. osmType(NODE, WAY) and then osmType(NODE) would result in only nodes being used, right?
So it's basically an intersection yeah, but in my understanding not an 'and' operation :)

[the remark this answers is not in the document]
Comment (tyrasd) [reply oshdb-api/src/test/java/org/heigit/bigspatialdata/oshdb/api/tests/TestOSMDataFilters.java]: I adjusted the PR description. In the Changelog it was already described a bit more clearly: https://github.com/GIScience/oshdb/pull/157/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR6

**Comment by tyrasd** [comment]
thanks for the review @FabiKo117. The only reason why I was still reluctant to merge this is that it technically did break the way the method `.osmType` works compared to the current (v0.5.5) version.

But, on the other hand, according to our documentation (https://github.com/GIScience/oshdb/blob/master/documentation/manual/filters.md) the current (v0.5.5) code is actually wrong because it states the following which is currently not the case for the `osmType` filter:

> Multiple filters can be applied after each other. The result will then contain any OSM elements that match all of the specified filters.

I'm changing this to a _bugfix_ (from a _breaking_ change), if you're ok with that?!

**Comment by FabiKo117** [comment]
I see.. from that point of view it is then actually a bugfix indeed.

**Pull request merged**